created api for app

CeNTePeDe opened the pull request.

CeNTePeDe committed:
fixed permissions:
- add method has_permission
added DefaultRouter
deleted file urls.py
created serializers
created views
created urls
created Swagger template
fixed files Pipfile and Pipfile.lock
created permissions:
- permission for customer, provider, car_showroom
fixed models:
- added connection OneToOne with user_model

slavadan commented on show_room_app/user/managers.py:
```diff
@@ -0,0 +1,47 @@
+            extra_fields.get("is_customer") is not True
+            or extra_fields.get("is_car_showroom") is not True
+            or extra_fields.get("is_provider") is not True
+        ):
```
not extra_fields.get("is_car_showroom")

slavadan commented on show_room_app/user/managers.py:
```diff
@@ -0,0 +1,47 @@
+        if extra_fields.get("is_staff") is not True:
+            raise ValueError(_("Superuser must have is_staff=True."))
+        if extra_fields.get("is_superuser") is not True:
+            raise ValueError(_("Superuser must have is_superuser=True."))
```
not extra_fields.get("is_staff")

slavadan commented on show_room_app/provider/views.py:
```diff
@@ -1,3 +1,45 @@
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        provider = get_object_or_404(queryset, pk=pk)
+        serializer = ProviderDetailSerializer(provider)
```
ты сделала `serializer_class`, а толку от него, если ты делаешь:  `serializer = ProviderDetailSerializer(provider)`

slavadan replied:
не гибко получается

slavadan commented on show_room_app/provider/models.py:
```diff
@@ -5,6 +5,7 @@
 
 class Provider(BaseRole):
     cars = models.ManyToManyField("cars.Car", through="CarProvider")
+    user = models.OneToOneField("user.User", on_delete=models.CASCADE, primary_key=True)
```
related_name прокинь

slavadan commented on show_room_app/discount/views.py:
```diff
@@ -1,3 +1,68 @@
+):
+    serializer_class = CarShowRoomDiscountSerializer
+    queryset = SeasonDiscount.objects.all()
+    permission_classes = [IsAdminUserOrReadOnly]
```
а @action так и не сделала((((

slavadan commented on show_room_app/discount/views.py:
```diff
@@ -1,3 +1,68 @@
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        season_discount = get_object_or_404(queryset, pk=pk)
+        serializer = SeasonDiscountSerializer(season_discount)
```
ты сделала serializer_class, а толку от него

slavadan replied:
Ты не поправила. Когда ты делаешь `serializer = SeasonDiscountSerializer(season_discount)` смысл от `serializer_class = SeasonDiscountSerializer` теряется. Те тут нужно `serializer = serializer_class(season_discount)`.
В других местах тоже поменяй

slavadan commented on show_room_app/customer/views.py:
```diff
@@ -1,3 +1,44 @@
+    def list(self, request, *args, **kwargs):
+        queryset = self.get_queryset()
+        serializer = CustomerSerializer(queryset, many=True)
+        return Response(serializer.data)
```
а чего ты list переопределила?

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    def has_object_permission(self, request, view, obj):
+        if obj.user == request.user and request.user.is_customer:
+            return True
+        return False
```
можно сделать одной строкой

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    """Custom permission for owner Car-Showroom."""
+
+    def has_permission(self, request, view):
+        return bool(request.method in permissions.SAFE_METHODS or request.user and request.user.is_staff)
```
а bool тут зачем?

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    def has_object_permission(self, request, view, obj):
+        if obj.user == request.user and request.user.is_car_showroom:
+            return True
+        return False
```
можно сделать одной строкой

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    """Custom permission for provider."""
+
+    def has_permission(self, request, view):
+        return bool(request.method in permissions.SAFE_METHODS or request.user and request.user.is_staff)
```
bool тут не нужен

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    def has_object_permission(self, request, view, obj):
+        if obj.user == request.user and request.user.is_provider:
+            return True
+        return False
```
одной строкой можно

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -0,0 +1,30 @@
+    is_car_showroom = serializers.BooleanField(required=False)
+
+    password1 = serializers.CharField(required=True, write_only=True)
+    password2 = serializers.CharField(required=True, write_only=True)
```
password1 password2??

никогда не используй цифры в названии переменных

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -0,0 +1,38 @@
+    """Custom permission for customer."""
+
+    def has_permission(self, request, view):
+        return bool(request.method in permissions.SAFE_METHODS or request.user and request.user.is_staff)
```
bool не нужен

slavadan commented on show_room_app/provider/views.py:
```diff
@@ -1,3 +1,45 @@
+
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        provider = get_object_or_404(queryset, pk=pk)
```
ты же вкурсе, что `get_object_or_404` делает raise? И если не обрабатывать его твое приложение ляжет. Когда объекта нету, просто возвращай 404 в респосе

slavadan commented on show_room_app/provider/views.py:
```diff
@@ -1,3 +1,45 @@
+
+    def list(self, request, *args, **kwargs):
+        queryset = self.get_queryset()
+        serializer = ProviderSerializer(queryset, many=True)
```
у тебя есть serializer_class

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,3 +1,41 @@
+
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        car_showroom = get_object_or_404(queryset, pk=pk)
```
читай ниже

slavadan commented on show_room_app/car_showroom/models.py:
```diff
@@ -7,6 +7,7 @@
 class CarShowRoom(BaseRole):
     cars = models.ManyToManyField("cars.Car", through="SellModel")
     balance = MoneyField(max_digits=14, decimal_places=2, default_currency="USD")
+    user = models.OneToOneField("user.User", on_delete=models.CASCADE, primary_key=True)
```
прокинь related_name или как там его

slavadan commented:
Добавь тайпхинты в функции

CeNTePeDe committed:
fixed views:
-optimized request
fixed permission:
-remove bool
added debug-toolbar
fixed serializer:
- rename serializer
- removed CarShowRoomSerializer, CarListSerializer, CustomerSerializer, ProviderSerializer,
fixed ViewSet:
- add delete methods
- override retrieve method
fix models:
-added related name for fields with OneToOne relations

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,41 +1,67 @@
+):
+    serializer_class = CarShowRoomSerializer
+    queryset = (
+        CarShowRoom.objects.filter(is_active=True)
+        .prefetch_related("cars")
+        .prefetch_related(
+            Prefetch('user', queryset=User.objects.all().only('username'))
+        )
+    )
```
Что то сложно. Почему ты два раза prefetch_related вызываешь вместо, того чтобы  через запятую в одном прокинуть.

slavadan replied:
так же еще можно переопределить get_queryset или как там его. Тож можешь потрогать

slavadan replied:
Отмечу, что если ты никак не дергаешь связанную модель(что-то наподобии model1.model2.name = 'anime'), то смысла делать префетчи и селекты нету.

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,41 +1,67 @@
+class SellModelView(viewsets.ModelViewSet):
     serializer_class = SellModelSerializer
-    queryset = SellModel.objects.all()
+    queryset = (
+        SellModel.objects.prefetch_related("discount")
+        .prefetch_related("season_discount")
+        .prefetch_related(
+            Prefetch('car_showroom', queryset=CarShowRoom.objects.all().only('name'))
+        )
+        .prefetch_related(
+            Prefetch('provider', queryset=Provider.objects.all().only('name'))
+        )
+        .prefetch_related(
+            Prefetch('car', queryset=Car.objects.all().only('name'))
+        )
+        .all()
+    )
```
емае

slavadan replied:
Ну ты нашаманила)))

slavadan replied:
словил передоз префетчей

slavadan replied:
а зачем .all в конце

slavadan commented on show_room_app/cars/views.py:
```diff
@@ -1,33 +1,50 @@
-        queryset = self.get_queryset()
-        serializer = CarListSerializer(queryset, many=True)
-        return Response(serializer.data)
+    http_method_names = ["get", "post", "patch", "delete"]
```
прикольно, не знал, что так можно

slavadan commented on show_room_app/cars/views.py:
```diff
@@ -1,33 +1,50 @@
+        permission_classes=[IsAdminUser],
+        url_path=r"",
+    )
+    def unpublished_cars(self, request):
+        unpublished = Car.objects.filter(is_active=False)
```
не понял, а почему unpublished_cars?

slavadan replied:
есть какие-то published cars? чето не то с неймингом

slavadan replied:
достаточно будет get_cars или car_list

CeNTePeDe replied:
Это я репетировала action, это только для админа неактивные машины, те они в базе есть , но не опубликованы

slavadan replied:
так и не переименовала

slavadan commented on show_room_app/customer/views.py:
```diff
@@ -15,30 +21,52 @@ class CustomerView(
     viewsets.GenericViewSet,
 ):
-    queryset = Transaction.objects.all()
+    queryset = Transaction.objects.prefetch_related(
+        Prefetch('car_showroom', queryset=CarShowRoom.objects.all().only('name'))
+    ).prefetch_related(
+        Prefetch('season_discount', queryset=SeasonDiscount.objects.all().only('discount_name'))
+    ).prefetch_related(
+        Prefetch('discount', queryset=CarShowRoomDiscount.objects.all().only('discount_name'))).all()
```
больше префетчей - богу префетчей

slavadan commented on show_room_app/discount/views.py:
```diff
@@ -54,6 +44,12 @@ class ProviderDiscountView(
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        provider_discount = get_object_or_404(queryset, pk=pk)
+        serializer = ProviderDiscountSerializer(provider_discount)
```
уже писал выше

slavadan commented on show_room_app/discount/views.py:
```diff
@@ -66,3 +62,9 @@ class CarShowRoomDiscountView(
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        car_showroom_discount = get_object_or_404(queryset, pk=pk)
+        serializer = ProviderDiscountSerializer(car_showroom_discount)
```
уже писал выше

slavadan commented on show_room_app/provider/views.py:
```diff
@@ -1,45 +1,58 @@
+        .prefetch_related(
+            Prefetch("user", queryset=User.objects.all().only("username"))
+        )
+        .filter(is_active=True)
```
уххх опять префетчи

slavadan commented on show_room_app/provider/views.py:
```diff
@@ -1,45 +1,58 @@
     serializer_class = CarProviderSerializer
-    queryset = CarProvider.objects.all()
+    queryset = (
+        CarProvider.objects.prefetch_related(
+            Prefetch('provider', queryset=Provider.objects.all().only('name'))
+        ).prefetch_related(
+            Prefetch('car', queryset=Car.objects.all().only('name'))
+        ).all()
```
префетчи

slavadan commented on show_room_app/show_room_app/settings.py:
```diff
@@ -126,38 +128,45 @@
     # ],
 }
+
+INTERNAL_IPS = [
+    os.environ.get("ALLOWED_HOST"),
+]
+
+if DEBUG:
+    import socket
+
+    hostname, _, ips = socket.gethostbyname_ex(socket.gethostname())
+    INTERNAL_IPS = [ip[: ip.rfind(".")] + ".1" for ip in ips] + [
+        ALLOWED_HOSTS,
+        "10.0.2.2",
+    ]
```
Впервые такое вижу

slavadan replied:
пояснительную бригаду пожалуйста

slavadan commented on show_room_app/show_room_app/urls.py:
```diff
@@ -10,25 +10,28 @@
+router.register(r"sell_car", SellModelView, basename="sell_car")
 router.register(r"customer", CustomerView, basename="customer")
 router.register(r"transaction", TransactionView, basename="transaction")
 router.register(r"season_discount", SeasonDiscountView, basename="season_discount")
```
Роуты кстати можно по апкам разбить и сюда докидывать. Если не поняла пиши в тг

slavadan replied:
посмотри какие роуты кроме DefaultRouter есть

slavadan commented on show_room_app/user/managers.py:
```diff
@@ -23,9 +23,9 @@ def create_user(
-            extra_fields.get("is_customer") is not True
-            or extra_fields.get("is_car_showroom") is not True
-            or extra_fields.get("is_provider") is not True
+            not extra_fields.get("is_customer") is not True
+            or not extra_fields.get("is_car_showroom") is not True
+            or not extra_fields.get("is_provider") is not True
```
тип если не False/True это не True. Каламбур какой-то. Вместо `not extra_fields.get("is_customer") is not True` можно просто `not extra_fields.get("is_customer")` ну или если прям хочется `extra_fields.get("is_customer") is False` ну или `extra_fields.get("is_customer") == False`. Если что-то не поняла насчет `not extra_fields.get("is_customer")`. И почитай какие есть варианты conditions, а то не понятно и не читаемо.

slavadan commented on show_room_app/user/managers.py:
```diff
@@ -40,8 +40,8 @@ def create_superuser(self, email, username, **extra_fields):
         extra_fields.setdefault("is_active", True)
 
-        if extra_fields.get("is_staff") is not True:
+        if not extra_fields.get("is_staff") is not True:
```
тоже самое

slavadan commented on show_room_app/user/managers.py:
```diff
@@ -40,8 +40,8 @@ def create_superuser(self, email, username, **extra_fields):
+        if not extra_fields.get("is_staff") is not True:
             raise ValueError(_("Superuser must have is_staff=True."))
-        if extra_fields.get("is_superuser") is not True:
+        if not extra_fields.get("is_superuser") is not True:
```
тоже самое

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -1,38 +1,56 @@
-
-    def has_object_permission(self, request, view, obj):
-        if obj.user == request.user and request.user.is_customer:
+    def has_permission(self, request: Any, view: Any) -> bool:
+        return (
+            request.method in permissions.SAFE_METHODS
+            or request.user
+            and request.user.is_staff
+        )
```
это у тебя форматтер так сделал? если нет то сделай в одну линию

CeNTePeDe replied:
Форматтер возращает

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -1,38 +1,56 @@
+        )
+
+    def has_object_permission(self, request: Any, view: Any, obj: Any) -> bool:
+        if (
+            obj.user == request.user and request.user.is_customer
+        ) or request.user.is_superuser:
```
ты немного не поняла. зачем тут if? можно же просто `return (obj.user == request.user and request.user.is_customer) or request.user.is_superuser`. Насчет тайпхинтов: request вообще-то имеет тип HttpRequest.

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -1,38 +1,56 @@
+            obj.user == request.user and request.user.is_car_showroom
+        ) or request.user.is_superuser:
             return True
-        return False
```
аналогично

slavadan commented on show_room_app/user/permission.py:
```diff
@@ -1,38 +1,56 @@
-
-    def has_object_permission(self, request, view, obj):
-        if obj.user == request.user and request.user.is_provider:
+    def has_permission(self, request: Any, view: Any) -> bool:
+        return (
+            request.method in permissions.SAFE_METHODS
+            or request.user
+            and request.user.is_staff
+        )
+
+    def has_object_permission(self, request: Any, view: Any, obj: Any) -> bool:
+        if (
+            obj.user == request.user and request.user.is_provider
+        ) or request.user.is_superuser:
```
аналогично

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -12,11 +12,11 @@ class RegisterSerializer(serializers.Serializer):
 
-    password1 = serializers.CharField(required=True, write_only=True)
-    password2 = serializers.CharField(required=True, write_only=True)
+    password = serializers.CharField(required=True, write_only=True)
+    password_confirm = serializers.CharField(required=True, write_only=True)
```
другое дело)

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -26,5 +26,4 @@ class Meta:
             "is_customer",
             "is_provider",
             "is_car_showroom",
-            "is_active",
```
а чего ты is_active убрала?

slavadan commented on show_room_app/user/views.py:
```diff
@@ -1,17 +1,25 @@
+    def retrieve(self, request, pk=id, *args, **kwargs):
+        queryset = self.get_queryset()
+        car = get_object_or_404(queryset, pk=pk)
+        serializer = UserSerializer(car)
```
про сериализаторы писал выше.

CeNTePeDe replied:
Вы же со Всеволодом по этому поводу спорили, и сказали, что так и надо только импорт из generics импортировать

CeNTePeDe committed:
created pytest.ini
created tests for app cars
updated admin pannel:
- added list_display
updated urls:
- removed path
updated CarShowRoom:
- created create method
update serializers:
- add RegisterSerializer for user
fixed views:
- removed object Prefetch
- import get_object_or_404 from generics
- created views foe userRegistration, create car_showroom
fixed urls:
-create urls for each apps

slavadan commented on show_room_app/car_showroom/models.py:
```diff
@@ -7,6 +7,12 @@
+        "user.User",
+        on_delete=models.CASCADE,
+        primary_key=True,
+        related_name="CarShowRoom",
```
Не заметил. CarShowRoom это будет название переменной у класса юзер. А как ты понимаешь CarShowRoom в таком стиле переменные не называют. CarShowRoom -> car_showroom

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,3 +1,69 @@
+    serializer_class = CarShowRoomSerializer
+    queryset = (
+        CarShowRoom.objects.filter(is_active=True)
+        .prefetch_related("cars")
+        .prefetch_related(
+            Prefetch("user", queryset=User.objects.all().only("username"))
+        )
```
два префетча

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,3 +1,69 @@
+
+class SellModelView(viewsets.ModelViewSet):
+    serializer_class = SellModelSerializer
+    queryset = SellModel.objects.prefetch_related(
+        "discount", "season_discount", "car_showroom", "provider", "car"
+    ).all()
```
все еще не понимаю зачем тебе .all

slavadan commented on show_room_app/cars/permission.py:
```diff
@@ -0,0 +1,12 @@
+    """
+
+    def has_permission(self, request, view):
+        return bool(
+            request.method in SAFE_METHODS or request.user and request.user.is_staff
+        )
```
не исправила

slavadan commented on show_room_app/provider/urls.py:
```diff
@@ -0,0 +1,12 @@
+router.register(r"provider", ProviderView, basename="provider")
+router.register(r"car_provider", CarProviderListView, basename="car_provider")
+
+urlpatterns = [
+    path("", include(router.urls)),
+]
```
я мб уже не помню, но вроде можно делать так: urlpatterns = router.urls
если можно то поменяй тут и в других местах

CeNTePeDe committed:
created factories for apps
created api_tests for discount, cars models
added moduls:
- pytest
- pytest-django
- faker
- faker-vehicle
- pytest-drf
- factory-boy
- pytest-factoryboy
deleted file tests

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+        "price": car.price.amount,
+        "is_active": car.is_active,
+    }
+    print("expected_json=", expected_json)
```
принтов в тестах не должно быть

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+from cars.serializer import CarSerializer
+
+ENDPOINT = "/api/v1_car/cars/"
+pytestmark = pytest.mark.django_db
```
переделай с использование декоратора

slavadan replied:
в других местах тоже

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def test_retrieve(client):
+    car = CarFakeFactory.create()
```
Ты тестируешь и API и сериализатор для Car в одном файле и при этом у тебя файл называется test_serializer???????!!!!! Я же вроде показывал и рассказывал про нейминг и структуру тестов по принципу
один файл - одна сущность. Те у тебя папка tests в ней __init__.py, conftest.py, factory.py
У тебя есть модель Car. значит файлик будет называться test_car.py. Также можно еще больше разделить в том плане чтобы не тестировать все сразу в одном файле можно сделать папку test_api например и в ней будет test_car.py и тестить апи для модели Car в этом файле, а для другого например test_model и в ней тоже будет test_car.py и тестировать модели или какую нибудь бизнес логику связанную с моделями car(но это не обязательно - главное чтобы ты смысл поняла).

slavadan commented on show_room_app/cars/tests/factories.py:
```diff
@@ -0,0 +1,45 @@
+
+from cars.choice import BodyType, Color, EnginType, NumberOfDoor
+from cars.models import Car
+from faker import Faker
+from faker_vehicle import VehicleProvider
+
+fake = Faker()
+fake.add_provider(VehicleProvider)
```
если ты используешь factory-boy, то и используй пожайлуста faker который он предоставляет)

slavadan replied:
переделай в других местах тоже

slavadan commented on show_room_app/cars/tests/factories.py:
```diff
@@ -0,0 +1,45 @@
+    engine_type = random.choice(EnginType.choices[1])
+    number_of_doors = random.choice(NumberOfDoor.choices[1])
+    price = Decimal("23.99")
+    is_active = True
```
в чем прикол двух фабрик для машины. Одной мало?

slavadan replied:
мощности фабрики не хватает чтоле Xd?

slavadan commented on show_room_app/cars/tests/test_model.py:
```diff
@@ -0,0 +1,21 @@
+    assert car.engine_type == "Hybrid"
+    assert car.number_of_doors == "five-door"
+    assert car.price.amount == Decimal("9.99")
+    assert car.is_active == True
```
на фикстуры тесты не пишутся

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+    assert json.loads(response.content)["body_type"] == car.body_type
+    assert json.loads(response.content)["engine_type"] == car.engine_type
+    assert json.loads(response.content)["number_of_doors"] == car.number_of_doors
+    assert json.loads(response.content)["price"] == str(car.price.amount)
```
Ну емае как так можно. Я очень сильно расстроен. Как думаешь почему?

slavadan replied:
жду ответ в тг

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+    assert json.loads(response.content)["body_type"] == car.body_type
+    assert json.loads(response.content)["engine_type"] == car.engine_type
+    assert json.loads(response.content)["number_of_doors"] == car.number_of_doors
+    assert json.loads(response.content)["price"] == str(car.price.amount)
```
Моя печаль не знает границ

slavadan commented on show_room_app/cars/tests/test_serializer.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def test_create_car(client):
+    car = CarFakeFactory.build()
```
Посмотри как делают фикстуры как фабричный метод и сделай также. В других местах тоже

slavadan replied:
Можно конечно и так оставить, но тогда ты не узнаешь что из фикстуры можно устроить фабричный метод, в которую можно доп логику прописывать. Мб покажу завтра пример

slavadan commented on show_room_app/discount/tests/conftest.py:
```diff
@@ -0,0 +1,18 @@
+
+register(SeasonDiscountFactory)
+register(ProviderDiscountFactory)
+register(CarShowRoomDiscountFactory)
```
кстати я не помню зачем нужен register. Но если сделаешь фабричный метод то register тут не нужен будет

slavadan commented on show_room_app/discount/tests/test_provider_discount.py:
```diff
@@ -0,0 +1,74 @@
+    )
+    print(new_provider_discount)
+    assert response.status_code == 200
+    assert json.loads(response.content) == new_provider_discount
```
Выглядит вроде неплохо, но есть принт в печаль в одном месте

slavadan commented on show_room_app/discount/tests/test_car_showroom_discount.py:
```diff
@@ -0,0 +1,76 @@
+        url,
+        new_car_showroom_discount,
+    )
+    print(new_car_showroom_discount)
```
принты

slavadan commented on show_room_app/user/tests/test_user.py:
```diff
@@ -0,0 +1,24 @@
+        "is_car_showroom": False,
+    }
+    response = client.post(url, payload)
+    data = response.data
```
опа, а тут я удивлен. А че раньше так нельзя было?

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,3 +1,59 @@
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(status=status.HTTP_403_FORBIDDEN)
```
у тебя же пермишен стоит. зачем тут if

CeNTePeDe committed:
added tests:
- test_delete_car
- test_delete_car_showroom
- test_delete_provider
- tests for sell_model
created fixture for
- sell_model
- customer
created tests for customer, transaction
created conftest.py for car_showroom, cars, discount, provider, customer
removed serializer for user
added decorator for tests func
created tests for apps
created factories for apps

CeNTePeDe commented on show_room_app/cars/tests/test_car.py:
```diff
@@ -67,20 +68,23 @@ def test_retrieve(client):
-    old_car = CarFakeFactory.create()
-    new_car = CarFakeFactory.build()
+@pytest.mark.django_db
+def test_update_car(client):
```
Все тесты на  update не срабатывают, выскакивает ошибка 405

CeNTePeDe commented on show_room_app/car_showroom/tests/test_car_showroom.py:
```diff
@@ -0,0 +1,47 @@
+    print(expected_json)
+    response = client.post(ENDPOINT, data=expected_json)
+    assert response.status_code == 201
+
```
тут с этим тоже проблемы, badrequest((((

slavadan commented on show_room_app/car_showroom/views.py:
```diff
@@ -1,3 +1,59 @@
+    #         serializer.is_valid(raise_exception=True)
+    #         serializer.save()
+    #         return Response(serializer.data, status=status.HTTP_201_CREATED)
+    #     return Response(status=status.HTTP_403_FORBIDDEN)
```
?

CeNTePeDe committed:
add api_client
add path for auth_change_password
added id in factory:
-CarFactory
-CarShowRoomFactory
created tests for apps
updated serializers
created fixtures apps

slavadan commented on show_room_app/car_showroom/tests/conftest.py:
```diff
@@ -0,0 +1,65 @@
+            season_discount=season_discount,
+            **kwargs
+        )
+
```
я вроде говорил перенести переменные в kwargs функции build. в других местах тоже

slavadan commented on show_room_app/car_showroom/tests/factories.py:
```diff
@@ -0,0 +1,56 @@
+from provider.tests.factories import ProviderFactory
+from user.tests.factories import UserFactory
+
+faker = Faker()
```
Про попробовать использовать Facker из factory-boy я тоже говорил. И не нужно будет объявлять переменные faker = Faker().
https://besnard-consulting.com/articles/django-testing/3-generating-random-datasets-with-factory-boy

в других местах тоже

slavadan commented on show_room_app/car_showroom/tests/test_car_showroom.py:
```diff
@@ -0,0 +1,81 @@
+        "user": car_showroom.user.id,
+        "is_active": car_showroom.is_active,
+    }
+    print(payload)
```
не убрала print

slavadan commented on show_room_app/cars/tests/test_car.py:
```diff
@@ -0,0 +1,116 @@
+    assert payload["engine_type"] == data["engine_type"]
+    assert payload["number_of_doors"] == data["number_of_doors"]
+    assert payload["price"] == Decimal(data["price"])
+    assert payload["is_active"] == data["is_active"]
```
Кстати на одном из ревью на моем проекте было рассказано про методологию написания тестов AAA(Arrange, Act, Assert). Почитай это очень интересно https://habr.com/ru/articles/507594/. И сделай также те Раздели отступами .(Arrange, Act, Assert). Если не понятно пиши в тг

slavadan commented on show_room_app/user/serializer.py:
```diff
@@ -0,0 +1,103 @@
+    password = serializers.CharField(
+        write_only=True, required=True, validators=[validate_password]
+    )
+    password2 = serializers.CharField(write_only=True, required=True)
```
случайно увидел, но про наличие цифр в названии переменных я уже говорил!!!

CeNTePeDe committed:
updated serializers for apps
craeted registration for user

CeNTePeDe committed:
fixed model transaction:
- add customer
fixed model customer
- remove column purchases
fixed tests:
- provider
- car_provider
- car_showroom
- customer
add to serializers method create
removed create_user, build_user from fixtures provider, car_showroom, customer
added:
 - migrations for apps

CeNTePeDe commented on show_room_app/car_showroom/serializer.py:
```diff
@@ -26,13 +26,22 @@ class Meta:
+    # season_discount = SeasonDiscountSerializer()
+    # provider = ProviderSerializer()
+    # car = CarSerializer()
 
```
смотри, может здесь все таки по id , просто модель связующая, и информация только для провайдера, а уже какие модели машины, провайдер может и запросом провести

CeNTePeDe replied:
ну и sellmodel по такому же принципу?

CeNTePeDe commented on show_room_app/car_showroom/tests/conftest.py:
```diff
@@ -7,19 +7,17 @@
 @pytest.fixture
-def build_car_showroom(create_user_car_showroom):
+def build_car_showroom():
     def car_showroom(**kwargs):
```
вот здесь убрала create_user, тк мы прокидываем в ручную

CeNTePeDe commented on show_room_app/customer/models.py:
```diff
@@ -27,6 +26,7 @@ class Transaction(models.Model):
         "car_showroom.CarShowRoom", on_delete=models.CASCADE
     )
+    customer = models.ForeignKey("customer.Customer", on_delete=models.CASCADE)
     price = MoneyField(max_digits=14, decimal_places=2, default_currency="USD")
```
сustomer я все таки добавила в transaction

CeNTePeDe replied:
transaction test еще не переделывала

CeNTePeDe replied:
я бы еще кинула может fk на машину, чтобы знать что за она

CeNTePeDe committed:
added filtres for cars
override fixtures
remove Faker to factory.Faker()
added create method for SellModel
added transaction field to Customer
added car field to Transaction
created create method for sell_model

CeNTePeDe commented on show_room_app/car_showroom/serializer.py:
```diff
@@ -54,3 +57,32 @@ class Meta:
+    #                                           season_discount=season_discount,
+    #                                           **validated_data)
+    #     return sell_model
+    #
```
Метод оставлен на всякий случай, )))

CeNTePeDe replied:
Вторая модель сделана по id

CeNTePeDe commented on show_room_app/customer/models.py:
```diff
@@ -8,8 +8,13 @@ class Customer(models.Model):
     max_price = MoneyField(max_digits=14, decimal_places=2, default_currency="USD")
+    transaction = models.ForeignKey(
+        "Transaction", on_delete=models.CASCADE, blank=True, null=True
+    )
```
я, конечно, кинула связь на transaction, НО как все таки работает FK , с какой стороны MANY а с какой ONE, по такой логике и на CarShowRoom надо разворачивать связь

CeNTePeDe commented on show_room_app/customer/tests/test_transaction.py:
```diff
@@ -15,13 +16,15 @@ def test_create_transaction(api_client, build_transaction):
     response = api_client.post(
         ENDPOINT, data=json.dumps(payload), content_type="application/json"
     )
+
```
Вот здесь по ID, просто по отдельности мы же проверяем (создание машины, custoner, car_showroom и тд), в этой модельке мы просто собираем все данные по существующим объектам и высчитываем цену автомобиля, на не надо их создавать мы просто их собираем вместе
